### DISCUSSION TITLE
Generating types gives unhelpful type definitions

### DISCUSSION THREAD

**Discussion by kopi-cloud** [supabase/supabase#344] [General]
I was very happy to see Supabase's support for generating types based on the schema definition described in the [doco](https://supabase.io/docs/client/generating-types).

But when I tried it out, the result was very disappointing.

Give a table defined like this:
```
create table public.welcome (
  id text constraint welcome_pkey primary key,
  value bigint not null
);
```

The end result is:
```
export interface definitions {
  ...
  /**
   * data for the welcome screen
   */
  welcome: { [key: string]: any };
}
```

I was expecting something more like:
```
export type welcome {
  id: string,
  value: number,
}
```

I don't know much about Swagger/OpenAPI, but it seems like the definition being generated from the `/rest/v1/` endpoint doesn't contain the structure of the table objects, for example the `welcome` definition is just:
```
    "welcome": {
      "description": "data for the welcome screen",
      "type": "object"
    }
```

Am I doing something wrong?  Is there some config or a flag I need to set to make the swagger endpoint generate more info?

I've [attached](https://github.com/supabase/supabase/files/5650767/swagger.json.zip) the full swagger document.

**Comment by talolard** [comment]
Hi,
Building on this and possibly hijacking (sorry if...) 

We're heavy postgrest users, and autogenerated and typed client libraries would be amazing. 
For our non postgrest needs we've used [openapi-client-axios](https://github.com/anttiviljami/openapi-client-axios) which generates type definitions and a typed client library. 
Unfortunately it works on openapi3 and postgrest is on openapi2 and getting a useful client has proven challenging. 

We've been thinking of paying the [maintainer](https://viljami.io/) of openapi-client-axios to make a similar library specifically for postgrest, would the Supabase team want to join ?

**Comment by kiwicopple** [comment]
Hey @talolard - the axios client looks interesting.

Our end goal for the type generation will skip the swagger/OpenApi completely though. Instead we will generate it straight from the database - there is too much information lost between the database and the OpenApi spec.

This is (one of) the reasons we're building up the [postgres-api](https://github.com/supabase/pg-api)

**Reply by kiwicopple** [reply]
We're hoping to have a better version of type gen released in Q1/Q2

**Reply by shorn** [reply]
So is the current swagger being generated as expected? (ie. no field/column definitions for the objects)

To be honest, I assumed the current behaviour was a bug or I was doing something wrong.

**Reply by kiwicopple** [reply]
actually after looking at @kopi-cloud 's zip, I think there is a bug


```json
"definitions": {
    "flyway_schema_history": {
      "type": "object"
    },
    "user_info": {
      "description": "info provided by the user",
      "type": "object"
    },
    "welcome": {
      "description": "data for the welcome screen",
      "type": "object"
    }
  },
```

When I run the "countries" QuickStart, my swagger definitions look like this:

```json
"definitions": {
    "countries": {
      "description": "Full list of countries.",
      "required": [
        "id",
        "iso2"
      ],
      "properties": {
        "id": {
          "format": "bigint",
          "type": "integer",
          "description": "Note:\nThis is a Primary Key.<pk/>"
        },
        "name": {
          "format": "text",
          "type": "string",
          "description": "Full country name."
        },
        "iso2": {
          "format": "text",
          "type": "string",
          "description": "ISO 3166-1 alpha-2 code."
        },
        "iso3": {
          "format": "text",
          "type": "string",
          "description": "ISO 3166-1 alpha-3 code."
        },
        "local_name": {
          "format": "text",
          "type": "string",
          "description": "Local variation of the name."
        },
        "continent": {
          "format": "public.continents",
          "type": "string",
          "enum": [
            "Africa",
            "Antarctica",
            "Asia",
            "Europe",
            "Oceania",
            "North America",
            "South America"
          ]
        }
      },
      "type": "object"
    },
}
```

@kopi-cloud - would you mind emailing us your URL and anon key? - beta@supabase.io

**Reply by kopi-cloud** [reply]
@kiwicopple sent the details.

I created the table "manual_for_kiwicopple" through the online editor thinking creating tables via direct SQL might be the issue, but the swagger url still doesn't have any field definitions.

**Reply by kopi-cloud** [reply]
@kiwicopple I just tried this in a different supabase project and it seems to be working properly there.
Have sent details of working project to the beta address.

**Reply by kopi-cloud** [reply]
Ok - it looks like this is my fault.

I'm using a product called Flyway to do schema migrations, and the supabase project that has problems is used as my continuous integration DB.  As such, I run the [flyway clean](https://flywaydb.org/documentation/usage/commandline/clean) command on it as part of my test pipeline.  I think the clean command has deleted objects in the public schema that supabase expects to be there.

Oops, my bad.

**Reply by kiwicopple** [reply]
No probs @kopi-cloud !

**Reply by kopi-cloud** [reply]
I used a brand new project to verify that my pipeline works properly from creating tables via migrations to generating swagger definitions through to generating types with proper field definitions - so it does work.

But, of course, I messed up my code and accidentally stomped the public schema again.  It would be convenient (if somewhat dangerous for a production dataset) if there was some way to reset the DB to a state of "empty  + supabase stuff".  Unless you can suggest an alternate approach, I'm going to have to create a new SB project and re-configure all the `postgres://` and `https://` config in dev each time I do that.

**Reply by kiwicopple** [reply]
We'll add Point in Time backups in the future - the easiest way is to do a pg_dump of your database before you start making changes, then if anything goes awry you can run a restore to overwrite your changes
